src/security/core/src/dds_security_timed_cb.c

```diff
@@ -0,0 +1,330 @@
+                                    event->dispatcher->listener,
+                                    event->arg);
+
+                    /* Remove handled event from queue, continue with next. */
+                    event = remove_event(event);
+                } else {
+                    /* Check next event. */
+                    event = event->next;
+                }
+            } else {
+                /* All following events are in the future.
+                 * Let the condition wait until that future is reached. */
+            }
+        } else {
+            /* No (more) events. Wait until some are added or until the end. */
+            timeout = DDS_INFINITY;
+        }
+
+        if ((timeout > 0) || (timeout == DDS_INFINITY)) {
+            /* Wait for new event, timeout or the end. */
+            (void)ddsrt_cond_waitfor(&g_cond, &g_lock, timeout);
+
+            /* Restart, because 'old' triggers could have been added. */
+            event = g_first_event;
+        }
+
+    } while (!g_terminate);
+
+    ddsrt_mutex_unlock(&g_lock);
+
+    return 0;
+}
+
+
+struct dds_security_timed_dispatcher_t*
+dds_security_timed_dispatcher_new()
+{
+    struct dds_security_timed_dispatcher_t *d;
+
+    /* Initialization. */
+    timed_cb_init();
+
+    /* New dispatcher. */
+    d = ddsrt_malloc(sizeof(struct dds_security_timed_dispatcher_t));
+    memset(d, 0, sizeof(struct dds_security_timed_dispatcher_t));
+
+    return d;
+}
+
+void
+dds_security_timed_dispatcher_free(
+        struct dds_security_timed_dispatcher_t *d)
+{
+    struct queue_event_t *event;
+
+    assert(d);
+
+    /* Remove d related events from queue. */
+    ddsrt_mutex_lock(&g_lock);
+    event = g_first_event;
+    while (event != NULL) {
+        if (event->dispatcher == d) {
+            event->callback(event->dispatcher,
+                            DDS_SECURITY_TIMED_CB_KIND_DELETE,
+                            NULL,
+                            event->arg);
+            event = remove_event(event);
+        } else {
+            event = event->next;
+        }
+    }
+    ddsrt_mutex_unlock(&g_lock);
+
+    timed_cb_fini();
+
+    /* Free this dispatcher. */
+    ddsrt_free(d);
+}
+
+
+void
+dds_security_timed_dispatcher_enable(
+        struct dds_security_timed_dispatcher_t *d, void *listener)
+{
+    assert(d);
+    assert(!(d->active));
+
+    dds_return_t osres;
+
+    ddsrt_mutex_lock(&g_lock);
+
+    /* Remember possible listener and activate. */
+    d->listener = listener;
+    d->active = true;
+
+    /* Start thread when not running, otherwise wake it up to
+     * trigger callbacks that were (possibly) previously added. */
+    if (g_thread_ptr == NULL ) {
+        g_thread_ptr = ddsrt_malloc(sizeof(*g_thread_ptr));
+        ddsrt_threadattr_t attr;
+        ddsrt_threadattr_init(&attr);
+        osres = ddsrt_thread_create(g_thread_ptr, "security_dispatcher", &attr, timed_dispatcher_thread, NULL);
+        if (osres == DDS_RETCODE_ERROR)
```

Comment:
The test for failure of thread creation should be `osres != DDS_RETCODE_OK` (e.g. the Windows abstration could return `DDS_RETCODE_OUT_OF_RESOURCES` because it allocates memory for the thread context)

Reply:
I have replaced the condition as suggested.